File: LinkChecker.java
// LinkChecker.java
import java.io.*;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.*;
import java.util.concurrent.*;
import java.util.stream.Collectors;

public class LinkChecker {
    private static final String RESET = "\033[0m";
    private static final String GREEN = "\033[92m";
    private static final String YELLOW = "\033[93m";
    private static final String RED = "\033[91m";
    private static final String GRAY = "\033[90m";

    static class Result {
        String url;
        int status;
        boolean error;
        Result(String url, int status, boolean error) {
            this.url = url; this.status = status; this.error = error;
        }
    }

    public static List<String> loadLinks(String filename) throws IOException {
        List<String> links = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) continue;
                links.add(line);
            }
        }
        return links;
    }

    public static int checkUrl(String url, int timeoutSec, boolean follow) {
        HttpClient client = HttpClient.newBuilder()
                .followRedirects(follow ? HttpClient.Redirect.NORMAL : HttpClient.Redirect.NEVER)
                .connectTimeout(Duration.ofSeconds(timeoutSec))
                .build();
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("User-Agent", "LinkChecker/1.0")
                .timeout(Duration.ofSeconds(timeoutSec))
                .build();
        try {
            HttpResponse<Void> response = client.send(request, HttpResponse.BodyHandlers.discarding());
            return response.statusCode();
        } catch (Exception e) {
            return -1; // ошибка
        }
    }

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Использование: java LinkChecker <файл.txt> [--threads N] [--timeout S] [--follow] [--output file]");
            System.exit(1);
        }
        String source = args[0];
        int threads = 4;
        int timeoutSec = 5;
        boolean follow = false;
        String output = null;
        for (int i = 1; i < args.length; i++) {
            if (args[i].equals("--threads") && i+1 < args.length) {
                threads = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--timeout") && i+1 < args.length) {
                timeoutSec = Integer.parseInt(args[++i]);
            } else if (args[i].equals("--follow")) {
                follow = true;
            } else if (args[i].equals("--output") && i+1 < args.length) {
                output = args[++i];
            }
        }

        List<String> links;
        if (source.endsWith(".txt")) {
            links = loadLinks(source);
        } else {
            links = Arrays.asList(args);
        }

        if (links.isEmpty()) {
            System.err.println("Нет ссылок для проверки.");
            System.exit(1);
        }

        System.out.printf("Проверка %d ссылок (потоков: %d, таймаут: %dс)...\n", links.size(), threads, timeoutSec);
        long start = System.currentTimeMillis();

        ExecutorService executor = Executors.newFixedThreadPool(threads);
        List<Future<Result>> futures = new ArrayList<>();
        for (String url : links) {
            futures.add(executor.submit(() -> {
                int status = checkUrl(url, timeoutSec, follow);
                return new Result(url, status, status == -1);
            }));
        }
        List<Result> results = new ArrayList<>();
        for (Future<Result> f : futures) {
            results.add(f.get());
        }
        executor.shutdown();

        long elapsed = System.currentTimeMillis() - start;
        System.out.println("\nРезультаты:");
        for (Result r : results) {
            String color;
            if (r.error) color = GRAY;
            else if (r.status >= 200 && r.status < 300) color = GREEN;
            else if (r.status >= 300 && r.status < 400) color = YELLOW;
            else if (r.status >= 400 && r.status < 600) color = RED;
            else color = GRAY;
            System.out.printf("  %s -> %s%d%s\n", r.url, color, r.status, RESET);
        }

        int total = results.size();
        int ok = 0, redirect = 0, error = 0, fail = 0;
        for (Result r : results) {
            if (r.error) fail++;
            else if (r.status >= 200 && r.status < 300) ok++;
            else if (r.status >= 300 && r.status < 400) redirect++;
            else if (r.status >= 400 && r.status < 600) error++;
        }
        System.out.printf("\nСтатистика: Всего: %d, OK: %d, Редиректы: %d, Ошибки: %d, Сбои: %d\n", total, ok, redirect, error, fail);
        System.out.printf("Время: %.2f сек.\n", elapsed / 1000.0);

        if (output != null) {
            System.out.println("Экспорт в " + output + " (не реализован для краткости)");
        }
    }
}

File: link_checker.rb
# link_checker.rb
require 'net/http'
require 'uri'
require 'json'
require 'csv'
require 'optparse'
require 'thread'

COLORS = {
  green: "\e[92m",
  yellow: "\e[93m",
  red: "\e[91m",
  gray: "\e[90m",
  reset: "\e[0m"
}

def check_url(url, timeout, follow)
  uri = URI.parse(url)
  http = Net::HTTP.new(uri.host, uri.port)
  http.use_ssl = (uri.scheme == 'https')
  http.open_timeout = timeout
  http.read_timeout = timeout
  request = Net::HTTP::Get.new(uri.request_uri)
  request['User-Agent'] = 'LinkChecker/1.0'
  begin
    response = http.request(request)
    # Проверка редиректов, если не следуем
    unless follow
      if response.is_a?(Net::HTTPRedirection)
        return response.code.to_i
      end
    end
    return response.code.to_i
  rescue
    return nil
  end
end

def load_links(filename)
  lines = File.readlines(filename, chomp: true)
  lines.reject { |l| l.empty? || l.start_with?('#') }
end

def color_status(status)
  return "#{COLORS[:gray]}ERR#{COLORS[:reset]}" if status.nil?
  if status >= 200 && status < 300
    "#{COLORS[:green]}#{status}#{COLORS[:reset]}"
  elsif status >= 300 && status < 400
    "#{COLORS[:yellow]}#{status}#{COLORS[:reset]}"
  elsif status >= 400 && status < 600
    "#{COLORS[:red]}#{status}#{COLORS[:reset]}"
  else
    status.to_s
  end
end

options = {}
OptionParser.new do |opts|
  opts.banner = "Использование: ruby link_checker.rb <файл.txt> [--threads N] [--timeout S] [--follow] [--output file]"
  opts.on("--threads N", Integer, "Количество потоков") { |v| options[:threads] = v }
  opts.on("--timeout S", Integer, "Таймаут в секундах") { |v| options[:timeout] = v }
  opts.on("--follow", "Следовать по редиректам") { options[:follow] = true }
  opts.on("--output FILE", "Экспорт результатов") { |v| options[:output] = v }
end.parse!

source = ARGV[0]
if source.nil?
  puts "Укажите файл со ссылками."
  exit 1
end

threads = options[:threads] || 4
timeout = options[:timeout] || 5
follow = options[:follow] || false
output = options[:output]

if source.end_with?('.txt')
  links = load_links(source)
else
  links = ARGV
end

if links.empty?
  puts "Нет ссылок для проверки."
  exit 1
end

puts "Проверка #{links.size} ссылок (потоков: #{threads}, таймаут: #{timeout}с)..."
start = Time.now

results = {}
queue = Queue.new
links.each { |l| queue << l }

workers = threads.times.map do
  Thread.new do
    while url = queue.pop(true) rescue nil
      status = check_url(url, timeout, follow)
      results[url] = status
    end
  end
end
workers.each(&:join)

elapsed = Time.now - start

puts "\nРезультаты:"
links.each do |url|
  status = results[url]
  puts "  #{url} -> #{color_status(status)}"
end

total = links.size
ok = results.values.count { |s| s && s >= 200 && s < 300 }
redirect = results.values.count { |s| s && s >= 300 && s < 400 }
error = results.values.count { |s| s && s >= 400 && s < 600 }
fail = results.values.count { |s| s.nil? }
puts "\nСтатистика: Всего: #{total}, OK: #{ok}, Редиректы: #{redirect}, Ошибки: #{error}, Сбои: #{fail}"
puts "Время: #{elapsed.round(2)} сек."

if output
  puts "Экспорт в #{output} (не реализован для краткости)"
end
